Sequent Calculus, Signed Modal-Logic Tableaux and Statistics

huterguier opened the pull request:
Sequenzenkalkül für Propositional- und First-Order-Logic.

rec0de commented:
Hi!
First off, thanks a lot for your work - I just tried it and everything seems to work very well.
I'll have a look at the calculus code itself in a bit, but for now here's a couple cleanup-things:

- There are a bunch of mostly empty files in `backend/.gradle` and `frontend/.gradle` that probably shouldn't be there, please remove those
- `sequentCalculus/psc/BasicMoves.kt` has no content and should be removed
- jacoco is imported twice in `build.gradle.kts` (once as `id("jacoco")` and once as `jacoco`
- I _think_ the abstract logic classes in `abstractLogic.kt` don't need to / shouldn't provide default implementations for `synEq

Also, in preparation for publishing your calculus at kbar.app, it'd be great if you could come up with one or two examples for the propositional / first order sequent`calculus. See for example "Stranded in Shanghai" https://kbar.app/fo-tableaux

huterguier commented:
Hello,
we have fixed all of the issues you mentioned except for the empty files in backend/.gradle and frontend/.gradle.  We dont know which files we are supposed to delete because we didnt change anything in the gradle settings and as far as we know all the files are created automatically by gradle.

As far as we know the examples have to be set in admin mode which we cannot access yet.  So here the two examples we have come up with:

Propositional:

It is warm in the summer.
summer & (summer -> sun) & (sun -> warm) |- warm

First Order:

There is a Mouse.
\all X: (Jerry(X) -> IsMouse(X)), \ex X: (\ex Y: (Jerry(X) & Tom(Y)) |- \ex X: (IsMouse(X))

rec0de commented:
Great, thanks! You can safely delete all the gradle files (from git), they are generated automatically and don't have to be in the repo.

huterguier commented:
We just deleted the .gradle folders. Thanks for the quick answer and the suggestions. We still have a few  weeks left and also a user study coming up. Were probably going to react to the feedback of the user study first but then we can also take a look at your suggestions.

DerDrodt commented:
> Either way, I'll probably merge this once you deleted the extra files - except if @DerDrodt or @JuliusHenke want to look over the frontend changes?

Oh, I’ll definitely look over them in detail :) I only skimmed them so far, but everything looks good code-wise. Unfortunately I did not have time to test it yet. 

It looks like very little of the structure had to be changed, especially for the frontend—which is hopefully a good sign. If you @huterguier et al. have some time, I would love to hear how you found working on the project and if the code base was as extendable as we planned. 

And I think it's based if we wait to merge until the project is actually finished. Just in case you change significant things.

rec0de commented:
Alright, then we'll leave this open for now

huterguier commented:
We thought we might be able to use the kbar.app website for our user study because its easier for the participants to reach it. Our other option would be the vm on which the gitlab-runner is located but that would be only reachable from within the uni network. So if you don't want to merge it already that's not a big problem but we would really appreciate if you did.

rec0de commented:
Ah, I see - we can probably make that work. When's your user study?

rec0de commented:
Alright, so:  
I manually deployed the new backend to api.kbar.app for now, which should be backwards-compatible with the current frontend.  
I also created a preview of the new frontend at https://kalkulierbar-8eda1--pr10-study-beta-ghwz92z1.web.app/ - which is not a pretty link by any measure, but I guess you can use a link shortener or something?  
If that works for you, I'd keep this pull request open as a draft for the final merge.  

Note the preview expires on March 4th.

JuliusHenke commented on frontend/src/components/calculus/sequent/node/index.tsx:
```diff
@@ -0,0 +1,284 @@
+
+        let lastMove: string = "";
+
+        if (ruleName === "notRight") {
+            lastMove = "¬R";
+        } else if (ruleName === "notLeft") {
+            lastMove = "¬L";
+        } else if (ruleName === "andRight") {
+            lastMove = "∧R";
+        } else if (ruleName === "andLeft") {
+            lastMove = "∧L";
+        } else if (ruleName === "orRight") {
+            lastMove = "∨R";
+        } else if (ruleName === "orLeft") {
+            lastMove = "∨L";
+        } else if (ruleName === "impLeft") {
+            lastMove = "→L";
+        } else if (ruleName === "impRight") {
+            lastMove = "→R";
+        } else if (ruleName === "Ax") {
+            lastMove = "Ax";
+        } else if (ruleName === "exLeft") {
+            lastMove = "∃L";
+        } else if (ruleName === "exRight") {
+            lastMove = "∃R";
+        } else if (ruleName === "allLeft") {
+            lastMove = "∀L";
+        } else if (ruleName === "allRight") {
+            lastMove = "∀R";
+        }
```
This could be turned into a switch case statement or a lookup table by Object literal. It would improve maintainability and readability. https://medium.com/front-end-weekly/switch-case-if-else-or-a-lookup-map-a-study-case-de1c801d944

JuliusHenke reviewed (commented):
Thanks for you work. It already looks really good. I only found one code quality issue in the frontend part.

JuliusHenke commented:
> Propositional:
> 
> It is warm in the summer.
> summer & (summer -> sun) & (sun -> warm) |- warm
> 
> First Order:
> 
> There is a Mouse.
> \all X: (Jerry(X) -> IsMouse(X)), \ex X: (\ex Y: (Jerry(X) & Tom(Y)) |- \ex X: (IsMouse(X))

@huterguier I already added your Propositional example to https://kalkulierbar-8eda1--pr10-study-beta-ghwz92z1.web.app/ but the  First Order one returns an error from the backend: `Expected ')' but got end of input`
Let me know when you fixed the backend parsing or the example and I will try to add it again :)

rec0de commented:
@JuliusHenke i think it's just missing a another `) ` after `Tom(Y))`

JuliusHenke commented:
@rec0de Ahh yes, ok I added the example `\all X: (Jerry(X) -> IsMouse(X)), \ex X: (\ex Y: (Jerry(X) & Tom(Y))) |- \ex X: (IsMouse(X))`

huterguier commented:
Thanks a lot for creating the temporary server and also for fixing the little syntax error we had in our example. The user study should take place earlier next week so it shouldn't be a problem that the preview expires next Thursday. I also fixed the code quality issue in the latest commit. In case u find any other issues regarding the new calculi or the code quality just let us know.

DerDrodt commented on frontend/src/util/sequent.ts:
```diff
@@ -0,0 +1,156 @@
+): Tree<SequentTreeLayoutNode> => {
+    const n = nodes[i];
+
+    if (n == null)
```
When does this happen? Prefer `===` over `==`.

DerDrodt reviewed (changes requested):
I was a bit pedantic, sorry. Most things a very minor, though some really must be fixed, in my opinion.

Some general things: Aim for consistency of folder names as well. E.g.: `frontend/src/components/svg/{SmallRec|horizontalList}`. They should be in kebab-case.

Note: I only checked the frontend.

huterguier commented:
Hi,
we have finished our "Bachelor Praktikum" and changed this pull request to also include the latest features we implemented.
We applied the feedback of our user study to the sequent-calculi. Furthermore we also added a signed tableaux for the modal logic. We also added a way to save statistics on proof completion. To keep it modular we added a StatisticCalculus Interface, which any Calculus can implement. For now we only added the Statistics for SequentCalculus and ModalLogicTableaux.

rec0de commented:
Great, congrats on finishing your project!
We'll look over the changes and let you know if there's anything that needs fixing over the next few days.

JuliusHenke commented on frontend/src/types/calculus/index.ts:
```diff
@@ -26,17 +36,21 @@ export enum Calculus {
     foResolution = "fo-resolution",
     ncTableaux = "nc-tableaux",
     dpll = "dpll",
+    psc = "psc",
```
I agree, since prop-sequent is much easier to read than the short abbreviation. New developers or your future self will thank you for using self explaining variable names.

JuliusHenke reviewed (changes requested):
Thanks a lot for your great work. 👍🔥
I added some suggestions for code quality improvements. I am pretty sure our original code also lacks some of the proposed improvements in some places. 😄

rec0de reviewed (commented):
Alright, I looked over the backend code - most of the comments are small style changes and tweaks that I don't really want to bother you with after you already finished the project. I suspect things are similar for the frontend changes.  
I'd suggest we merge this into a local branch where we can tweak whatever we want to tweak and then take it live from there.

The only major issue I see is with the score / leaderboard feature - I think it's a really cool feature and I'd like to have it in the final product in some form, but the current implementation is vulnerable to SQL injection attacks in several places, potentially letting anyone take over the database and, with some luck / skill, perhaps even the entire server. I'm also not super hyped to add sqlite as a backend dependency.  

So for now, I'd say we disable / remove the scoring feature and if anyone has the time & inspiration to find a better way, we can add it back later (maybe as part of statekeeper and in-memory only? wouldn't be too bad if highscores get lost on a server restart, would it?)

JuliusHenke commented:
@rec0de I like the idea of a local branch where we can tweak the little stuff and also try out the Frontend with the preview URL.

I think highscores shouldn't be kept in memory but rather in a DB or state file, since I would be sad when my score gets lost on a server restart 😁 I agree with disabling the statistics feature until someone had time to secure it.

DerDrodt commented on frontend/src/components/dialog/save-stats-dialog/index.tsx:
```diff
@@ -0,0 +1,198 @@
+     * @returns {void}
+     */
+    const submit = () => {
+        const textInput = document.getElementById("name");
```
No direct DOM access.

DerDrodt commented on frontend/src/components/header/index.tsx:
```diff
@@ -18,27 +19,27 @@ interface HeaderProps {
-        </Fragment>
-    );
+    const right =
+        window.innerWidth < 950 ? (
```
If you change the header width, we probably should also adjust `smallScreen`.

DerDrodt commented on frontend/src/components/header/index.tsx:
```diff
@@ -59,7 +60,7 @@ const Header: preact.FunctionalComponent<HeaderProps> = ({ currentUrl }) => {
             <Dialog
                 class={style.dialog}
-                open={!smallScreen && open}
+                open={!(window.innerWidth < 900) && open}
```
Why 900?

DerDrodt commented on frontend/src/components/input/fab/check-close.tsx:
```diff
@@ -1,21 +1,32 @@
+    /**
+     * Function to call when the proof is valid
+     */
+    onProoven?: (stats: Statistics) => void;
```
`onProven`

DerDrodt commented on frontend/src/components/svg/horizontalList/index.tsx:
```diff
@@ -0,0 +1,275 @@
+                dims.x,
+                selectFormulaCallback,
+                selected,
+            ).map((el) => el)}
```
Why map `el` to `el`? And why not use a Fragment?

huterguier commented on backend/src/main/kotlin/kalkulierbar/signedtableaux/BasicMoves.kt:
```diff
@@ -0,0 +1,475 @@
+    if (!state.backtracking)
+        throw IllegalMove("Backtracking is not enabled for this proof")
+
+    state.moveHistory.add(Prune(nodeID))
```
As far as i can remember the move history is used for the undo move, at least in the signed tableaux.

rec0de replied:
You're right, I missed that
